### PR TITLE
Add logs agent pipeline performance telemetry 

### PR DESCRIPTION
**Pull request by gh123man**
### What does this PR do?

This PR introduces logs pipeline performance telemetry + makes internal buffers configurable. 

Pipeline telemetry is comprised of two major pieces: 

**Utilization Ratio** - Measurement of the ratio of how busy a unit of logic is vs it's idle time. 

**Utilization Items/Bytes** - measurement of the number of elements and bytes present in a single component (and it's input channel).

Both of these metrics are exposed as gauges. Due to how fast log messages are processed - the raw data is very noisy. So to mitigate this some simple aggregation is performed: 
- utilization ratio is summed up over a 1 second window and aggregated into an ewma which is exposed as a Gauge. 
- utilization items/bytes is measured by computing the ewma of `ingress - egress` every second and reporting it as a Gauge. 

In order to measure capacity we have to record ingress and egress throughout the logs pipeline. This means adding ingress and egress markers around when channel operations happen. 

This PR is broken down into individual commits that should make it easier to review. 
 
### Motivation

Increase visibility of logs agent performance to drive optimization efforts. 

### Describe how to test/QA your changes

- QA done by SMP run on PR
- Spot check check utilization still works

### CODE REVIEW

**Review by hush-hush — APPROVED** [review]
Looks good for ASC

File: pkg/logs/metrics/metrics.go
Comment (vickenty) [inline pkg/logs/metrics/metrics.go:88]: I fear the name of this metric may be misleading. Capacity is normally understood as maximum amount something can contain or handle, while this only reflects current usage level. How about `utilization.seconds`, `utilization.payloads`, `utilization.bytes`?

Also, would it be possible to place the explanation of this measures from the PR description here?

File: pkg/logs/processor/processor.go
Comment (vickenty) [inline pkg/logs/processor/processor.go:250]: Would it make sense to consider waiting on the downstream as also utilization time,  since we're not handling inbound messages?

File: pkg/logs/processor/processor.go
Comment (vickenty) [inline pkg/logs/processor/processor.go:256]: Should this and `.Stop` calls probably need to happen in early returns on error as well?

File: pkg/logs/metrics/metrics.go
Comment (gh123man) [reply pkg/logs/metrics/metrics.go:88]: That's fair. 
Using your suggestions i'd make the following tweaks:

Utilization is a ratio of in_use vs idle. So maybe `utilization.ratio`? (instead of `seconds`)
And then `utilization.payloads` should probably be `utilization.items` or `utilization.messages`? 
`utilization.bytes` sounds good to me. 

WDYT?

File: pkg/logs/processor/processor.go
Comment (gh123man) [reply pkg/logs/processor/processor.go:250]: This is something ive been back and forth on. 

If we include writing to the downstream channel as utilization, if a downstream component becomes a bottleneck (ex sender latency), it will cause all upstream utilization to go to `1` (100%) since it spends most of it's time blocking on the channel write. 

Whereas if we don't count the write to the downstream channel, the utilization stops at its current "maximum" when blocked by a downstream component. 

I opted for the second one since it makes the data easier to reason about (and more actionable if we were to create a monitor off of it in the future). The idea is to measure only the amount of "work" the component is doing, and blocking on a write is not really "work" IMO. 

(I have examples of both approaches, I can DM them to you if you are interested).

**Review by jade-guiton-dd — APPROVED** [review]
Looks good on the OTel side

File: pkg/logs/metrics/metrics.go
Comment (gh123man) [reply pkg/logs/metrics/metrics.go]: Ive made the above changes. LMK if you have further feedback.

**Review by blt — APPROVED** [review]

**Review by purple4reina — APPROVED** [review]

File: test/regression/cases/file_to_blackhole_1000ms_latency/experiment.yaml
Comment (goxberry) [inline test/regression/cases/file_to_blackhole_1000ms_latency/experiment.yaml:39]: Minor nit: would suggest using `0KiB` for consistency with the existing checks. The experiment configs look good to me. Thanks for adding these experiment configs!

**Review by goxberry — APPROVED** [review]

File: test/regression/cases/file_to_blackhole_1000ms_latency/experiment.yaml
Comment (gh123man) [reply test/regression/cases/file_to_blackhole_1000ms_latency/experiment.yaml]: Fixed!

**Review by remeh — COMMENTED** [review]
I have some concerns but there are some great ideas in there!

The main concern I that we do a lot of additions dispatched here and there where it's necessary, and that's what they actually are, additions, more than fitting in the overall design of the pipeline. It makes all of this a bit brittle and easily error-prone (e.g. forgetting to report an ingress), but it's probably not the best time to do such a bigger change neither and I'd still be OK shipping it that way first.

Another minor concern is that it seems that there is a mix of "declaring ingress entering the receiver" and "declaring ingress leaving the emitter", which is not ideal, but it might be caused by one of the mistake I found (I think).

File: pkg/logs/auditor/auditor.go
Comment (GustavoCaso) [inline pkg/logs/auditor/auditor.go:107]: Ideally we should provide the config component as part of `New` function, so we do not use the `pkgconfigsetup.Datadog()` global 😄 

Maybe something we could do on a separate PR

File: pkg/logs/diagnostic/message_receiver.go
Comment (GustavoCaso) [inline pkg/logs/diagnostic/message_receiver.go:52]: Same comments about using the global config accessor 😄

File: pkg/logs/metrics/metrics.go
Comment (GustavoCaso) [inline pkg/logs/metrics/metrics.go:93]: Same as the config global accessor, we should consider not using the telemetry global accessor. 

I know is a bit painful to refactor this parts of the codebase, but the more we keep using global the harder is going to be in the future. I'm happy to help with the effort.

File: pkg/logs/tailers/file/tailer.go
Comment (gh123man) [reply pkg/logs/tailers/file/tailer.go:347]: Oops yeah this should have been deleted.

File: pkg/logs/metrics/capacity_monitor.go
Comment (gh123man) [reply pkg/logs/metrics/capacity_monitor.go:66]: Yep I missed renaming this when adding the ewma

File: pkg/logs/tailers/file/tailer.go
Comment (gh123man) [reply pkg/logs/tailers/file/tailer.go:369]: WIll do. 
IMO the copy is fine - not ideal, but not hurting anything.

File: pkg/logs/pipeline/provider.go
Comment (gh123man) [reply pkg/logs/pipeline/provider.go:187]: I left the old call site around because other parts of the code base use the logs pipeline but not for logs (epforwarder). 

I should call this out in the PR description but this PR only attempts to measure the file tailer for now. 
I was planning to address the other cases in a followup PR once we get some real world testing on this data.

File: pkg/logs/processor/processor.go
Comment (gh123man) [reply pkg/logs/processor/processor.go:122]: Yes :( I missed this one as part of an earlier refactor.

File: pkg/logs/metrics/utilization_monitor.go
Comment (vickenty) [inline pkg/logs/metrics/utilization_monitor.go:48]: Not sure we need a dependency to do `old*(1-factor)+new*factor`, but if we use one: what is the factor here, how fast it reacts to changes? Is it going to be stable between library versions? Would it make sense to make it configurable?

File: pkg/logs/metrics/utilization_monitor.go
Comment (vickenty) [inline pkg/logs/metrics/utilization_monitor.go:67]: Only updating the utilization on `Stop`, means that if the pipeline is fully blocked for some reason (e.g. due to a bug downstream), the utilization metric will never be updated to reflect that. 

Also, if the pipeline is blocked repeatedly on a full 10s then the utilization metric will climb slower, than if the destination fails after being blocked for 5s (because we add fewer 1s to it with the longer timeout). This may be difficult to interpret correctly during troubleshooting.

Would it be possible to always update the utilization metric regardless of the state the pipeline is in? (For example, how we do for utilization tracking in the collector worker).

File: pkg/logs/processor/processor.go
Comment (vickenty) [reply pkg/logs/processor/processor.go:251]: Good point about pinpointing the bottlenecks. My angle was more that the .Stop() needs to be sprinkled just right to be correct, and it would be less error-prone to use `defer` so that happens automatically.

File: pkg/logs/processor/processor.go
Comment (vickenty) [reply pkg/logs/processor/processor.go:257]: I see you added .Stop(), but what about the `ReportComponentEgress`? If it is not called on all codepaths that forward or drop the message, then `utilization` just keeps growing.

File: pkg/logs/sender/sender.go
Comment (vickenty) [inline pkg/logs/sender/sender.go:136]: If we don't include channel wait time in the utilization, should this come before calls to `Send`?

File: pkg/logs/tailers/file/tailer.go
Comment (vickenty) [inline pkg/logs/tailers/file/tailer.go:367]: I think you leaving the following line as is probably answers the question remy had here quite positively. Would it make sense to remove this comment?

**Review by vickenty — COMMENTED** [review]
Would it make sense to add some tests for the new telemetry code?

File: pkg/logs/processor/processor.go
Comment (gh123man) [reply pkg/logs/processor/processor.go:251]: Yeah agreed. I haven't come up with a better solution yet to having to sprinkle the start/stop in just the right place.

File: pkg/logs/processor/processor.go
Comment (gh123man) [reply pkg/logs/processor/processor.go:257]: True. Ill defer `pipelineMonitor.ReportComponentEgress(msg, "processor")` in this case. 
These errors basically never happen (unless there is an underlying logic error) but it should be correct regardless.

File: pkg/logs/tailers/file/tailer.go
Comment (gh123man) [reply pkg/logs/tailers/file/tailer.go:367]: Ill defer to @remeh for this (I think it's fine to remove)

File: pkg/logs/metrics/utilization_monitor.go
Comment (gh123man) [reply pkg/logs/metrics/utilization_monitor.go:48]: > Not sure we need a dependency
It was already in our codebase - so I reached for it. Happy to reimplement it if preferred. 

> what is the factor here
The docs explain it [here](https://pkg.go.dev/github.com/VividCortex/ewma?utm_source=godoc#pkg-constants): `NewMovingAverage` uses the library defaults (one min window). 

>  Is it going to be stable between library versions?
[It hasn't changed in 11 years](https://github.com/VividCortex/ewma/blame/master/ewma.go#L7) - so seems pretty likely. 

> Would it make sense to make it configurable?
[It is via the constructor](https://github.com/VividCortex/ewma/blob/v1.2.0/ewma.go#L40) however the default seems good enough for now. 

These metrics are for internal use only at the moment to help us drive optimizations so we may choose to change the ewma settings in the future.

File: pkg/logs/sender/sender.go
Comment (gh123man) [reply pkg/logs/sender/sender.go:136]: So the sender is a bit of a special case. 
The sender doesn't do any actual work other than dispatch payloads to destinations - which do the work of actually sending them. 

So the "work" the sender does is is really the max of all destination utilization. For example - with dual shipping, if we are sending to US1 and EU1, and we become sender constrained because the EU1 has a longer round trip, the sender's utilization will increase despite the US1 destination utilization remaining low. 

I am actually not sure how useful this will be in practice - but I included it anyway as we can always remove it if we find otherwise.

File: pkg/logs/processor/processor.go
Comment (vickenty) [reply pkg/logs/processor/processor.go:251]: If `.Stop()` is changed to be idempotent, it should be possible to both `defer` it and call before channel write manually.

File: pkg/logs/tailers/file/tailer.go
Comment (remeh) [reply pkg/logs/tailers/file/tailer.go:367]: Please, feel free to remove it.
At that time I was curious that we were doing a copy instead of passing the data, but for sure this will work.

**Review by remeh — APPROVED** [review]
+1 on Vikentiy's comment about possibly adding unit tests on the new telemetry code if possible.

File: pkg/logs/metrics/utilization_monitor.go
Comment (gh123man) [reply pkg/logs/metrics/utilization_monitor.go]: Closing this since I removed this dep

File: pkg/logs/metrics/utilization_monitor.go
Comment (gh123man) [reply pkg/logs/metrics/utilization_monitor.go]: re-used our internal telemetry tracker 👍

File: pkg/logs/processor/processor.go
Comment (gh123man) [reply pkg/logs/processor/processor.go]: Addressed in https://github.com/DataDog/datadog-agent/pull/30744/commits/2b953ccb99df3e17e26b1f62d6030068d3dcbc1c

**Review by KevinFairise2 — APPROVED** [review]

**Review by CelianR — CHANGES_REQUESTED** [review]

**Review by CelianR — APPROVED** [review]

File: pkg/logs/metrics/capacity_monitor.go
Comment (vickenty) [inline pkg/logs/metrics/capacity_monitor.go:67]: We discussed this elsewhere already, but just for the record: a Histogram is probably more appropriate choice for a fast moving metric like this, since it will also capture short spikes that can be missed by periodic sampling like this, that would still nonetheless contribute to the agent's memory usage.

**Review by vickenty — APPROVED** [review]

**Pull request merged**
